lib/client/api.go

```diff
@@ -1897,54 +1928,105 @@ func (tc *TeleportClient) ConnectToProxy(ctx context.Context) (*ProxyClient, err
 // successful.
 func (tc *TeleportClient) connectToProxy(ctx context.Context) (*ProxyClient, error) {
 	proxyPrincipal := tc.getProxySSHPrincipal()
-	sshConfig := &ssh.ClientConfig{
-		User:            proxyPrincipal,
-		HostKeyCallback: tc.HostKeyCallback,
-	}
 
 	sshProxyAddr := tc.Config.SSHProxyAddr
 	if len(tc.JumpHosts) > 0 {
 		log.Debugf("Overriding SSH proxy to JumpHosts's address %q", tc.JumpHosts[0].Addr.String())
 		sshProxyAddr = tc.JumpHosts[0].Addr.Addr
+
+		// Perform a "dummy" connection to the jumphost to examine its certificate
+		// and load a corresponding user certificate to be used by tc.authMethods.
+		//
+		// All the loading logic happens in HostKeyCallback and it will always
+		// force the connection to fail.
+		//
+		// TODO(andrej): Get rid of the individually packaged authMethods,
+		// and embed this reissue logic with ssh.PublicKeysCallback.
+		tmpConfig := &ssh.ClientConfig{
+			User:            proxyPrincipal,
+			HostKeyCallback: tc.loadKeyForClusterFromCert(ctx),
+		}
+		// This connection will always fail so no need to capture return
+		// values.
+		ssh.Dial("tcp", sshProxyAddr, tmpConfig)
 	}
 
-	// helper to create a ProxyClient struct
-	makeProxyClient := func(sshClient *ssh.Client, m ssh.AuthMethod) *ProxyClient {
-		return &ProxyClient{
+	sshConfig := &ssh.ClientConfig{
+		User:            proxyPrincipal,
+		HostKeyCallback: tc.HostKeyCallback,
+	}
+	// try to authenticate using every non interactive auth method we have:
+	var errs []error
+	for i, m := range tc.authMethods() {
```

Comment:
here, `tc.authMethods()` will return `nil` if SSH cert is missing for the cluster.
`tc.authMethods` (and the underlying `LocalKeyAgent.AuthMethods()`) should return an error if no SSH cert is present.